feature_gates/__init__.py

```diff
@@ -0,0 +1,136 @@
+from dataclasses import dataclass
+from configparser import ConfigParser, NoOptionError, NoSectionError
+from typing import Final
+from pathlib import Path
+
+
+@dataclass
+class _FeatureGatesConfig:
+    """
+    Configuration for feature gates.
+
+    We follow the following convention:
+
+    feature_property: bool | None
+
+    If the value is None, the feature is not set.
+
+    To configure via `pyteal.ini`, the section should be `pyteal-feature`
+    with values such as
+    property = True
+
+    For example, the fields:
+
+    sourcemap_enabled: bool | None = None
+    sourcemap_debug: bool | None = None
+
+    imply a pyteal.ini file that looks something like:
+
+    [pyteal-sourcemap]
+    enabled = True
+    debug = False
+    """
+
+    sourcemap_enabled: bool | None = None
+    sourcemap_debug: bool | None = None
+
+
+class FeatureGates:
+    INI_FILENAME = "pyteal.ini"
+    INI_SECTION_PREFIX = "pyteal-"
+    DEFAULTS: Final[_FeatureGatesConfig] = _FeatureGatesConfig(
+        sourcemap_enabled=False,
+        sourcemap_debug=False,
+    )
+    _gates: Final[_FeatureGatesConfig] = _FeatureGatesConfig()
+    _features: Final[set[str]] = set(vars(_gates).keys())
+    verbose: bool = False
+
+    @classmethod
+    def _make_feature(cls, feature: str):
+        setattr(cls, feature, classmethod(lambda cls: cls.get(feature)))
+        setattr(
+            cls, f"set_{feature}", classmethod(lambda cls, gate: cls.set(feature, gate))
+        )
+
+    @classmethod
+    def _get(cls, feature: str) -> bool | None:
+        """Return the value of the given feature."""
+        if feature not in cls._features:
+            raise ValueError(
+                f"Unknown {feature=} (available features: {cls._features}))"
+            )
+
+        return getattr(cls._gates, feature)
+
+    @classmethod
+    def get(cls, feature: str) -> bool:
+        if (gate := cls._get(feature)) is None:
+            raise ValueError(f"Feature {feature} is not set")
+
+        return gate
+
+    @classmethod
+    def _set(cls, feature: str, gate: bool, overwrite: bool) -> None:
+        """Set the feature gate if it is unset, or overwrite == True."""
+        current = cls._get(feature)
+        if current is None or overwrite:
+            setattr(cls._gates, feature, gate)
+
+    @classmethod
+    def set(cls, feature: str, gate: bool) -> None:
+        """Set the feature gate for the given feature to the given value."""
+        cls._set(feature, gate, overwrite=True)
+
+    @classmethod
+    def load(cls, *, overwrite: bool = False, config_path: str = ""):
```

Comment:
I like the feature gates.  I hate the filesystem access.

Reply:
That means that the only way to change the default behavior of a feature gate is to explicitly call it in every indendent execution -for example- via `FeatureGate.set_sourcemap_enabled(False)`.

You wouldn't be able to just write the config file and have sourcemapping work universally in your project. 

Are you ok with that?

Reply:
I am ok with that.  Is @algoanne ok with that